### DISCUSSION TITLE
What is in the pipe for v3?

### DISCUSSION THREAD

**Discussion by leroyvn** [atmtools/arts#864] [General]
Hello! At this point, can you share some information about the major changes you are making for ARTS v3?

**Comment by riclarsson** [comment]
Hello Vincent,

The guiding principle for ARTS 3 design is that it should be easier to do things that are simple without losing out on more complicated physics.  In ARTS 2, we have a much more rigid system.  Of course, this puts quite some work on us to provide clean examples for how you do both the simple, and the difficult, calculations :)

There are going to be major changes throughout ARTS going from version 2 to version 3.  To list some:

1) We no longer have any main grids.  We have many smaller grids.
  1.a) All the different fields of the atmosphere in ARTS 2 must be along a `p_grid`, maybe a `lat_grid`, and yet maybe also a `lon_grid`.  You must have O2 VMR on the same grid as altitude on the same grid as temperature in ARTS 2.  In ARTS 3 the atmospheric elements each may have their own custom grid.  Technically, we make use of variant grids, so you can , e.g., use a constant O2 VMR regardless of at what position in the atmospheric field you sample for an atmospheric point, while temperature is just a 1D profile which only changes if you change the sampling altitude.  You can even pass in custom operators that allows you to compute the atmospheric field state as you sample the field.  
  1.b) Related to this change is that we consider altitude the vertical coordinate for sampling the atmospheric field in ARTS 3.
  1.c) If you work with retrievals, the grid of retrieval is now fixed to the atmospheric field parameters grid.  So if you want to retrieve a constant VMR, you just have a constant VMR in the atmospheric field.
2) The scattering interface is reworked from ground-up.  This is still ongoing work.  The main idea is to provide a more intuitive interface that allows running custom models of scattering more easily.  It should not anymore be required to massage your data to fit the ARTS 2 format, but instead you need to massage how this data is sampled.  Possibly by your own custom operator.
3) The python interface is much improved at the cost of having to scrap the `controlfile.arts` interface.  Some of these things like an improved GUI cf the thing I made that's in ARTS 2, and more direct ways to interface with libraries like xarray and the likes.
4) You are able to run ARTS 3 on Windows machines.
5) Spectral radiances are always fully polarised.
6) The path geometry of the atmosphere is always in full 3D.

There's probably a lot more but those are a few.

**Reply by leroyvn** [reply]
Thanks for this very detailed answer!

**Reply by leroyvn** [reply]
Sorry to reopen: I have another question.

Do you also foresee a release on PyPI for v3?

**Comment by riclarsson** [comment]
We will probably not be available on PyPI again.

ARTS is mainly a C++ software with python bindings.  This fits much better into the managed environments of miniforge3.  You can still easily install ARTS without compiling it yourself if you have miniforge3 via

ARTS 2.6:
```
mamba install rttools::pyarts
```

ARTS 3-dev:
```
mamba install rttools-dev::pyarts
```

I think they are supposed to be mutually exclusive by dependencies design, but they are anyways mutually exclusive by the fact both yields "import pyarts" and nothing else.

If you build ARTS 3 manually, the same method as for ARTS 2.6 can be used via "pip" to install it in your python environment.

Obligatory warning:

ARTS 3 is under heavy development.  Do not expect anything to work or be documented yet as even the glue it provides is still in flux.  It works only under managed situations (like we use it to teach our students) and not in scientific settings (because you will not be able to trace results back to a proper software version).
